mpas_analysis/shared/climatology/climatology.py

```diff
@@ -515,19 +530,105 @@ def update_start_end_year(ds, config, calendar):
     endYear = days_to_datetime(ds.Time.max().values,  calendar=calendar).year
     changed = False
     if startYear != requestedStartYear or endYear != requestedEndYear:
-        warn("climatology start and/or end year different from requested\n"
-             "requestd: {:04d}-{:04d}\n"
-             "actual:   {:04d}-{:04d}\n".format(requestedStartYear,
-                                                requestedEndYear,
-                                                startYear,
-                                                endYear))
+        warnings.warn("climatology start and/or end year different from requested\n"
+                      "requestd: {:04d}-{:04d}\n"
+                      "actual:   {:04d}-{:04d}\n".format(requestedStartYear, requestedEndYear, startYear, endYear))
         config.set('climatology', 'startYear', str(startYear))
         config.set('climatology', 'endYear', str(endYear))
         changed = True
 
     return changed, startYear, endYear
 
 
+def add_months_and_days_in_month(ds, calendar):
+    '''
+    Add ``months`` and ``daysInMonth`` as data arrays in ``ds``.  The number
+    of days in each month of ``ds`` is computed either using the ``startTime``
+    and ``endTime`` if available or assuming ``gregorian_noleap`` calendar and
+    ignoring leap years.
+
+    Parameters
+    ----------
+    ds : ``xarray.Dataset`` object
+        A data set with a ``Time`` coordinate expressed as days since
+        0001-01-01
+
+    calendar: ``{'gregorian', 'gregorian_noleap'}``
+        The name of one of the calendars supported by MPAS cores, used to
+        determine ``month`` from ``Time`` coordinate
+
+    Returns
+    -------
+    ds : ``xarray.Dataset`` object
+        The data set with ``month`` and ``daysInMonth`` data arrays added (if
+        not already present)
+
+    Authors
+    -------
+    Xylar Asay-Davis
+
+    Last Modified
+    -------------
+    03/29/2017
+    """    '''
+
+    ds = ds.copy()
+
+    if 'month' not in ds.data_vars and 'month' not in ds.coords:
+        if calendar is None:
+            raise ValueError('calendar must be provided if month coordinate is not in ds')
+        months = [date.month for date in days_to_datetime(ds.Time,
+                                                          calendar=calendar)]
+
+        ds['month'] = ('Time', months)
+
+    if 'daysInMonth' not in ds.data_vars:
+        if 'startTime' in ds.data_vars and 'endTime' in ds.data_vars:
+            ds['daysInMonth'] = ds.endTime - ds.startTime
+        else:
+            # TODO: support calendar
+            daysInMonth = numpy.array([constants.daysInMonth[month-1] for
+                                       month in ds.month.values], float)
+            ds['daysInMonth'] = ('Time', daysInMonth)
+
+    return ds
+
+
+def _compute_masked_mean(ds):
+    '''
+    Compute the time average of data set, masked out where the variables in ds
```

Comment:
So there are 2 things xarray's mean calculations don't seem to do that this is meant to handle.   First, there's not an obvious way to weight a mean by another variable (`daysInMonth` in this case).  Second, we want the normalization to be based on the number of *valid* (non-null) entries, so we need to also sum up how many valid entries there were and multiply them by the weights to determine how to do our weighted average.  For observational data sets, the mask changes with time, so it is very important to only average over those dates that are valid.  If one were to compute an annual mean using valid values from only Jan. and Feb. but were to divide by 365 days to get the weighted average, the resulting value would be off by a factor of about 6.  This function attempts to properly take that into account.

If you would like to rewrite the function to work in a more xarray-like way, feel free.  I have written many iterations of it and still haven't found a perfect solution.